pinot-core/src/main/java/org/apache/pinot/core/operator/query/LinearSelectionOrderByOperator.java

```diff
@@ -0,0 +1,452 @@
+   *   <li>If the new element belongs to the last included partition, the last partition is treated as a priority queue
+   *   sorted by the in-partition comparator. If the new element is lower than some of the already inserted elements,
+   *   the new replace the older.</li>
+   * </ul>
+   *
+   * This class can be used to implement order-by queries that include one or more not sorted expressions.
+   * In cases where all expressions are sorted, {@link TotallySortedListBuilder} should be used because its performance
+   * is better.
+   *
+   * As usual, elements are sorted by partition and there is no order guarantee inside each partition. The second
+   * comparator is only used to keep the lower elements in the last partition.
+   */
+  private static class PartiallySortedListBuilder implements ListBuilder {
+    /**
+     * A list with all the elements that have been already sorted.
+     */
+    private final ArrayList<Object[]> _sorted;
+    /**
+     * This attribute is used to store the last partition when the builder already contains {@link #_maxSize} elements.
+     */
+    private PriorityQueue<Object[]> _lastPartitionQueue = null;
+    /**
+     * The comparator that defines the partitions and the one that impose in which order add has to be called.
+     */
+    private final Comparator<Object[]> _partitionComparator;
+    /**
+     * The comparator that sorts different rows on each partition.
+     */
+    private final Comparator<Object[]> _unsortedComparator;
+    @Nullable
+    private Object[] _lastPartitionRow;
+    private final int _maxSize;
+    private int _alreadySorted;
+
+    public PartiallySortedListBuilder(int maxSize, Comparator<Object[]> partitionComparator,
+        Comparator<Object[]> unsortedComparator) {
+      Preconditions.checkArgument(maxSize > 0, "Max size must be greater than 0");
+      _maxSize = maxSize;
+      _sorted = new ArrayList<>(maxSize);
+      _partitionComparator = partitionComparator;
+      _unsortedComparator = unsortedComparator;
+    }
+
+    @Override
+    public boolean add(Object[] row) {
+      if (_lastPartitionRow == null) {
+        _lastPartitionRow = row;
+        _sorted.add(row);
+        return false;
+      }
+      int cmp = _partitionComparator.compare(row, _lastPartitionRow);
+      if (cmp < 0) {
+        throw new IllegalArgumentException(
+            "Row with docId " + _sorted.size() + " is not sorted compared to the previous one");
+      }
+
+      boolean newPartition = cmp > 0;
+      if (_sorted.size() < _maxSize) { // we don't have _maxSize elements yet
+        if (newPartition) {
+          _lastPartitionRow = row;
+          _alreadySorted = _sorted.size();
+        }
+        // just add the new row to the result list
+        _sorted.add(row);
+        return newPartition;
+      }
+      // at least _maxSize have been added
+      if (newPartition) { // and the new element belongs to a new partition, so we can just ignore it
+        return true;
```

Comment:
`_alreadySorted` is not updated, which can cause the while check for `sortedSize()` never able to reach